### PR TITLE
Move the project file handling for secret config to one place

### PR DESCRIPTION
**Pull request by Frassle**
# Description

A continuing clean up of secrets.

This means the secret providers just work in terms of workspace.ProjectStack, mutate as they wish and let the higher level work out if it should save the file or not. Rather than having each secret provider have to deal with loading and saving the stack config files all that that logic is now moved out to `getStackSecretsManager` and `createSecretsManager`.

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works - Should be covered by existing tests. No behavioral changes intended. 

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change - Not user facing

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by Frassle** [comment]
This broke some tests because of our cursed "only load the project files once into a global singleton" logic, fixed now but ugh that stuff's grim.

**Comment by RobbieMcKinstry** [comment]
> This broke some tests because of our cursed "only load the project files once into a global singleton" logic, fixed now but ugh that stuff's grim.

Wuhhh? Is this something we could refactor? I wouldn't mind doing it. That feels like it's going to be a blocker for some config changes I've been eyeing up for the middle-distance future.

**Comment by Frassle** [comment]
> Wuhhh? Is this something we could refactor? I wouldn't mind doing it. That feels like it's going to be a blocker for some config changes I've been eyeing up for the middle-distance future.

See https://github.com/pulumi/pulumi/blob/master/sdk/go/common/workspace/loaders.go which is full of singleton objects used to load various config files (Pulumi.yaml, Pulumi.<stack>.yaml, PulumiPlugin.yaml, etc)

I assume it was to try and save IO, but at the expense of global mutable state (everyone's favorite kind of state). All added in this PR: https://github.com/pulumi/pulumi/pull/6576

There's probably a sensible design to get the best of both worlds, but I think I'd be quite happy to just rip the whole thing out and refactor other bits of the codebase to not be hitting the working directory and loading files so much. Really all that ought to be happening like once at the cmd entry point and everything else through the system should be parameters.

Trying to clean up individual call sites to these load methods would be a good place to start, we can chip away at this it doesn't need to be a big bang clean up.

File: pkg/cmd/pulumi/crypto.go [hunk 1]
```diff
@@ -56,31 +57,50 @@ func getStackSecretsManager(s backend.Stack) (secrets.Manager, error) {
...
 	}
+
+	// Handle if the configuration changed any of EncryptedKey, etc
+	if err := saveProjectStackAfterSecretManger(s, oldConfig, ps); err != nil {
+		return nil, err
+	}
 	return stack.NewCachingSecretsManager(sm), nil
 }
 
+func saveProjectStackAfterSecretManger(stack backend.Stack,
```
Comment (abhinav) [inline pkg/cmd/pulumi/crypto.go:84]: Deep copy and compare on just those fields feels like it'll break easily in the future.

This is probably fine right now, but would a longer term fix here be if the individual secrets manager constructors did one of the following?

- return whether they modified ProjectStack (e.g. `func NewCloudSecretsMangaer(...) (sm secrets.Manager,  changed bool, err error)`
- return ProjectStackSecretSettings (or something) that the caller merges into ProjectStack
- mark the ProjectStack dirty with a `func (*ProjectStack) SetDirty(bool)` where they normally would have saved, and the caller inspects it (`func (*ProjectStack) Dirty() bool`) to decide if it needs to save

**Review by abhinav — APPROVED** [review]

File: pkg/cmd/pulumi/crypto.go (hunk 1 above)
Comment (Frassle) [reply pkg/cmd/pulumi/crypto.go:84]: > return ProjectStackSecretSettings (or something) that the caller merges into ProjectStack

I'm planning on doing something like that. This current system clearly won't work if we have user defined plugin secret systems.

**Comment by Frassle** [comment]
bors merge

**Pull request merged**